core/lib/libtomcrypt/src/tee_ltc_provider.c

```diff
@@ -2546,293 +2532,288 @@ struct tee_ccm_state {
-		if (ltc_res != CRYPT_OK)
-			return TEE_ERROR_BAD_STATE;
+	/* Add the IV */
+	ltc_res = ccm_add_nonce(&ccm->ctx, nonce, nonce_len);
+	if (ltc_res != CRYPT_OK)
+		return TEE_ERROR_BAD_STATE;
 
-		/* Add the IV */
-		ltc_res = gcm_add_iv(&gcm->ctx, nonce, nonce_len);
-		if (ltc_res != CRYPT_OK)
-			return TEE_ERROR_BAD_STATE;
-		break;
-#endif
-	default:
-		return TEE_ERROR_NOT_SUPPORTED;
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_aes_ccm_update_aad(void *ctx, const uint8_t *data, size_t len)
+{
+	struct tee_ccm_state *ccm = ctx;
+	int ltc_res;
+
+	/* Add the AAD (note: aad can be NULL if aadlen == 0) */
+	ltc_res = ccm_add_aad(&ccm->ctx, data, len);
+	if (ltc_res != CRYPT_OK)
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_aes_ccm_update_payload(void *ctx, TEE_OperationMode mode,
+					 const uint8_t *src_data,
+					 size_t len, uint8_t *dst_data)
+{
+	int ltc_res, dir;
+	struct tee_ccm_state *ccm = ctx;
+	unsigned char *pt, *ct;	/* the plain and the cipher text */
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		pt = (unsigned char *)src_data;
+		ct = dst_data;
+		dir = CCM_ENCRYPT;
+	} else {
+		pt = dst_data;
+		ct = (unsigned char *)src_data;
+		dir = CCM_DECRYPT;
 	}
+	ltc_res = ccm_process(&ccm->ctx, pt, len, ct, dir);
+	if (ltc_res != CRYPT_OK)
+		return TEE_ERROR_BAD_STATE;
 
 	return TEE_SUCCESS;
 }
 
-static TEE_Result authenc_update_aad(void *ctx, uint32_t algo,
-				     TEE_OperationMode mode __unused,
-				     const uint8_t *data, size_t len)
+TEE_Result crypto_aes_ccm_enc_final(void *ctx, const uint8_t *src_data,
+				    size_t len, uint8_t *dst_data,
+				    uint8_t *dst_tag, size_t *dst_tag_len)
 {
-#if defined(CFG_CRYPTO_CCM)
-	struct tee_ccm_state *ccm;
-#endif
-#if defined(CFG_CRYPTO_GCM)
-	struct tee_gcm_state *gcm;
-#endif
+	TEE_Result res;
+	struct tee_ccm_state *ccm = ctx;
 	int ltc_res;
 
-	switch (algo) {
-#if defined(CFG_CRYPTO_CCM)
-	case TEE_ALG_AES_CCM:
-		/* Add the AAD (note: aad can be NULL if aadlen == 0) */
-		ccm = ctx;
-		ltc_res = ccm_add_aad(&ccm->ctx, data, len);
-		if (ltc_res != CRYPT_OK)
-			return TEE_ERROR_BAD_STATE;
-		break;
-#endif
+	/* Finalize the remaining buffer */
+	res = crypto_aes_ccm_update_payload(ctx, TEE_MODE_ENCRYPT, src_data,
+					    len, dst_data);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* Check the tag length */
+	if (*dst_tag_len < ccm->tag_len) {
+		*dst_tag_len = ccm->tag_len;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+	*dst_tag_len = ccm->tag_len;
+
+	/* Compute the tag */
+	ltc_res = ccm_done(&ccm->ctx, dst_tag,
+			   (unsigned long *)dst_tag_len);
+	if (ltc_res != CRYPT_OK)
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_aes_ccm_dec_final(void *ctx, const uint8_t *src_data,
+				    size_t len, uint8_t *dst_data,
+				    const uint8_t *tag, size_t tag_len)
+{
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_ccm_state *ccm = ctx;
+	int ltc_res;
+	uint8_t dst_tag[TEE_xCM_TAG_MAX_LENGTH];
```

Comment:
minor: s/`TEE_xCM_TAG_MAX_LENGTH`/`TEE_CCM_TAG_MAX_LENGTH`/ here and line 2665.

(definition of `TEE_xCM_TAG_MAX_LENGTH` could be removed.